### ISSUE TITLE
[Finder] Initial directory is opened twice

### ISSUE THREAD

**Issue by mvorisek** [symfony/symfony#50851]
### Symfony version(s) affected

any / latest

### Description

in https://github.com/phpstan/phpstan-src/commit/da56f5bcdbadf6487f034ed87db4c720c8ee1095#diff-d2a02971e55cfbdc9f9994aea2bebd84108382ab5628b662d1f35ec2a70fbe3fR206 you can see the individual FS calls done by Symfony Finder

currently, it opens the initial directory twice, which costs some extra CPU time

### How to reproduce

see https://github.com/phpstan/phpstan-src/commit/f3b86c3ccad58ecf0289efded5485a70c20a6d8a#diff-87cf61100cce8074792c993360180be397a8325612856338be61c41eeb60389dR83

basically simple and typical usage:
```
<?php declare(strict_types = 1);

use Symfony\Component\Finder\Finder;

$path = __DIR__ . '/some_dir';
$fileExtensions = ['php', 'phpt'];

$finder = (new Finder())
    ->in($path)
    ->followLinks()
    ->files()
    ->name('*.{' . implode(',', $fileExtensions) . '}');
```

### Additional Context

IO is especially slow on Windows

**Comment by mvorisek** [comment]
This issue is caused by https://github.com/symfony/symfony/pull/8120 , more precisely: https://github.com/symfony/symfony/blame/v6.3.1/src/Symfony/Component/Finder/Iterator/RecursiveDirectoryIterator.php#L136
